enhancements/disaster-recovery/automated-cluster-backups.md

```diff
@@ -0,0 +1,324 @@
+      scheduled backup for (TODO How long)
+  - If the number of successful backups matches `retentionCount`
+    - Ensure removal of the oldest backup from the node that it was written to
+    - Remove record of removed backup from the api?
+    - Log removal of the backup
+  - Find the set of candidate nodes that have fewer backups than other nodes
+    - Could be 1 node (e.g. 2 nodes have 1 backup and 1 node has zero)
+    - Could be a subset of nodes (e.g. 1 node has 1 backup and 2 nodes have
+      zero)
+    - Could be all nodes (i.e. no successful backups have been taken)
+  - Schedule a backup pod to a random member of the set of candidate nodes
+    - A backup pod should invoke the `cluster-backup.sh` script on the host
+      via chroot and write the backup data to the host.
+    - A backup pod should first check that the available disk space on the
+      node it is running on is a multiple of the size of the node's
+      /var/lib/etcd path. A node needs a minimum amount of available storage
+      to operate reliably and automated backup must make every effort not to
+      exhaust a node's available storage.
+  - When a backup pod terminates, record the succcess or failure to a backup
+    resource
+
+### User Stories
+
+- I want to be alerted after installation or upgrade to the availability of
+  an automated backup capability and be given an indication of the importance
+  of configuring it in a timely manner.
+
+- I want to initiate a cluster backup without requiring a root shell on a
+  control plane node so as to minimize the risk involved.
+
+- I want to schedule recurring cluster backups so that I have recent cluster
+  state to recover with in the event of disaster (i.e. losing 2 control plane
+  nodes).
+
+- I want to have failure to take cluster backups for more than a configurable
+  period to be reported to me via critical alerts.
+
+- I want to minimize the operational impact of taking regular backups.
+
+### Risks and Mitigations
+
+Spreading the spread the cost of automated backups over all control plane
+nodes may not be sufficient to avoid negatively impacting cluster performance
+on IO-constrained hosts. Ideally it would be possible to limit the IO
+overhead of taking a backup via cgroup limits, but it's not clear that this
+would be achievable without the granularity provided by cgroups v2 (not
+slated for GA until RHEL9).
+
+Automated backups relying on host storage could fill up the disks of control
+plane nodes, rendering them inoperable. This is intended to be mitigated by a
+check for available disk space that ensures backups are only initiated on
+nodes that have a multiple of the space required to backup the node's
+/var/lib/etcd path.
+
+Relying on node storage may hamper recovery if all control plane nodes are
+lost. It might make sense to document the requirement to periodically ship
+backup data off the cluster. Given that the current restore procedure
```

Comment:
Should we make an off-cluster schedule shipper (e.g. S3 or similar...) part of this enhancement?

Reply:
The initial scoping is explicitly targeting an MVP that is generally useful for all clusters, not just those that have the ability to utilize off-cluster storage.